build: add and use C_STANDARD and CXX_STANDARD in depends

fanquake opened the pull request:
By explicitly setting a C standard version we avoid any potential for issues/differences in libraries that may come about due to C STD version, as well as avoid potentially being opted into newer code / features in libraries when compiler defaults change (i.e as of 11.0.0, Clang now defaults to gnu17 over gnu11).

This should be a no-op for our release builds, because it's just explicitly setting the default that is [already being used](https://github.com/fanquake/core-review/blob/master/compiler-defaults.md). However this is relevant for anyone building depends with a newer compiler.

I found [one broken `__STDC_VERSION__` check in the](https://github.com/miniupnp/miniupnp/pull/552) miniupnpc header.

At the same time, add `CXX_STANDARD` for setting our C++ standard, and use that over setting `-std=c++17` for cxx packages.

Guix builds:
```bash
```

hebasto commented:
Concept ACK.

> This should be a no-op for our release builds...

Could be verified on by-package basis with #21995.

laanwj commented:
Concept ACK.

Would it make sense to define a global e.g. `EXTRA_CFLAGS` variable and use it for every package, instead of repeating this for every single one?

tryphe commented:
Concept ACK

Zero-1729 commented:
Concept ACK

> Would it make sense to define a global e.g. `EXTRA_CFLAGS` variable and use it for every package, instead of repeating this for every single one?

Is there a reason for not going this route?

fanquake committed:
build: add and use C_STANDARD in depends

fanquake commented:
I've reworked this, and based it on top of #22840.

> Would it make sense to define a global e.g. EXTRA_CFLAGS variable and use it for every package, instead of repeating this for every single one?

I've added `DEFAULT_C_STD` and `DEFAULT_CXX_STD` variables to the depends makefile. Defaulting to `c11` and `c++17`. This means the versions are set in one place, and can also be overriden if someone wants to experiment building with a different version. i.e `gmake -C  depends DEFAULT_CXX_STD=c++20`.

jaysonmald35 commented:
Thank. U

fanquake commented:
Rebased now that #22840 is in. Would be good to get feedback on the new approach.

fanquake commented:
Guix Builds:
```bash
bash-5.1# find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
ba6a9c335b06f7b9afe8ea1513fc63f3d428ea241da62d6846b11b261dd3a2f5  guix-build-e5e8749daeb5/output/aarch64-linux-gnu/SHA256SUMS.part
d93909a10e7acdf9a63a1040403b52e2ac3cb8a85e012e8f0bb78fe00839f6a6  guix-build-e5e8749daeb5/output/aarch64-linux-gnu/bitcoin-e5e8749daeb5-aarch64-linux-gnu-debug.tar.gz
1d138c883bf71c3574892819143ef4a2c9c731be8a1366ea233d15dbe9d90eb5  guix-build-e5e8749daeb5/output/aarch64-linux-gnu/bitcoin-e5e8749daeb5-aarch64-linux-gnu.tar.gz
b2b958ba4bfa75a21c7a84c6b4d1644ca7205d9267ced9a912c3221fc1d664f3  guix-build-e5e8749daeb5/output/arm-linux-gnueabihf/SHA256SUMS.part
0fd5a404255c518a35af40077f2034ec00d1ad4c8eee622f4909c79714de68b7  guix-build-e5e8749daeb5/output/arm-linux-gnueabihf/bitcoin-e5e8749daeb5-arm-linux-gnueabihf-debug.tar.gz
5c0bdf64a39d6ba37f2443ec499fa6c294e5133df3d800f60862d49d28d2b88e  guix-build-e5e8749daeb5/output/arm-linux-gnueabihf/bitcoin-e5e8749daeb5-arm-linux-gnueabihf.tar.gz
59f206c960deb9c733ae4d4a676af82476f339d63bd776605d5fc0bb8e1b8ac6  guix-build-e5e8749daeb5/output/dist-archive/bitcoin-e5e8749daeb5.tar.gz
542143b7131b10649b76998ea1de4cdbdfd9b796e4a31ca2c78db7d9db37bef0  guix-build-e5e8749daeb5/output/powerpc64-linux-gnu/SHA256SUMS.part
41db837f0c5f6d3097d315612e0255f4f783b13264ad5d8eae8c7b7f1a153926  guix-build-e5e8749daeb5/output/powerpc64-linux-gnu/bitcoin-e5e8749daeb5-powerpc64-linux-gnu-debug.tar.gz
c989df06f68d111b3198b74e613fbdbe71a2aec97c186d4a3f90af23ffd4542b  guix-build-e5e8749daeb5/output/powerpc64-linux-gnu/bitcoin-e5e8749daeb5-powerpc64-linux-gnu.tar.gz
4b1d8912c434bd83d587ea9a29babe2954c102eca9c4c37500bb7e2f8f3b98fc  guix-build-e5e8749daeb5/output/powerpc64le-linux-gnu/SHA256SUMS.part
b081e53627bce90f92ca727f480872afebbb4f7048253bcec786659e5cc77309  guix-build-e5e8749daeb5/output/powerpc64le-linux-gnu/bitcoin-e5e8749daeb5-powerpc64le-linux-gnu-debug.tar.gz
fb9ea6e6fc1b464d92227acdbe2ff2a402f443203d4887de61f7adb3b9e9a5c5  guix-build-e5e8749daeb5/output/powerpc64le-linux-gnu/bitcoin-e5e8749daeb5-powerpc64le-linux-gnu.tar.gz
f88148d0493c8378fd529d85c4cbe886f3107ecfb633232db3ae48f13c18273e  guix-build-e5e8749daeb5/output/riscv64-linux-gnu/SHA256SUMS.part
00e818e80a3ec86a0b22c9f4008430b7cbf42224c53aac3ee712f9e60131385f  guix-build-e5e8749daeb5/output/riscv64-linux-gnu/bitcoin-e5e8749daeb5-riscv64-linux-gnu-debug.tar.gz
565d23085df0209c9cab686224bec4b75785d38a29a21983ec21564a4182eb8c  guix-build-e5e8749daeb5/output/riscv64-linux-gnu/bitcoin-e5e8749daeb5-riscv64-linux-gnu.tar.gz
e134b2e233ee3f7d5ea74343132ead88899f19a0c19038fccff633d0ef367955  guix-build-e5e8749daeb5/output/x86_64-apple-darwin18/SHA256SUMS.part
b8728894dde57c2dd6f3c68cb0aa971edc1d943c75bdb93dd92030e94e4b89a9  guix-build-e5e8749daeb5/output/x86_64-apple-darwin18/bitcoin-e5e8749daeb5-osx-unsigned.dmg
8055876fe46779d358d3724aa2ec07666e2fcc6034c4210b40f577066d75d15b  guix-build-e5e8749daeb5/output/x86_64-apple-darwin18/bitcoin-e5e8749daeb5-osx-unsigned.tar.gz
15dd2c86e1f01642012ac4b988bb3900c76017e6df208f3a29069a2ffd2a8b65  guix-build-e5e8749daeb5/output/x86_64-apple-darwin18/bitcoin-e5e8749daeb5-osx64.tar.gz
0f21a0c4c13e33252e3fdae3526b01d5364ab4be0daf7c9a984273720487258f  guix-build-e5e8749daeb5/output/x86_64-linux-gnu/SHA256SUMS.part
22f3cc5c6cb7b1e0f95314bd9690fe60e2bed1a3f7810ea8f41045479ea71abb  guix-build-e5e8749daeb5/output/x86_64-linux-gnu/bitcoin-e5e8749daeb5-x86_64-linux-gnu-debug.tar.gz
5b9e1a657fcc8fafb9530bd12869ecbab8aad40619a430f87ac57e82c840328b  guix-build-e5e8749daeb5/output/x86_64-linux-gnu/bitcoin-e5e8749daeb5-x86_64-linux-gnu.tar.gz
8d0bb218a91b5d81c03fedbfec8675a1d435a183c85909fc6f773d5c49ee2bea  guix-build-e5e8749daeb5/output/x86_64-w64-mingw32/SHA256SUMS.part
85e17ab08dc0448ac52e99bdcda8fa256814d43063e04faa4fe9bdfaeb05973a  guix-build-e5e8749daeb5/output/x86_64-w64-mingw32/bitcoin-e5e8749daeb5-win-unsigned.tar.gz
5920f3eca897c5b6c74bf778a2b0098b20392d5ece0b30d60772c7622bd0a8c1  guix-build-e5e8749daeb5/output/x86_64-w64-mingw32/bitcoin-e5e8749daeb5-win64-debug.zip
45d75041c5a729838917bb7e08af5107ed15fb49c6c69afe12d65e864cafb177  guix-build-e5e8749daeb5/output/x86_64-w64-mingw32/bitcoin-e5e8749daeb5-win64-setup-unsigned.exe
2f7ce8f5338e85b0fadb9e32770f5b84e29b0725202477033276e1b36077e7c6  guix-build-e5e8749daeb5/output/x86_64-w64-mingw32/bitcoin-e5e8749daeb5-win64.zip
```

theuni commented:
Concept ACK. Definitely better to be explicit.

Nit: `DEFAULT_FOO` implies that it's the fallback unless something specific was requested, which is not the case here as the value can be overridden. Would prefer to drop/replace the `DEFAULT_` prefix.

dongcarl commented:
Concept ACK!

Wondering: instead of setting it for every package, could we do it in `hosts/*.mk`?

practicalswift commented:
Concept ACK

fanquake committed:
build: add and use CXX_STANDARD in depends

fanquake commented:
> Nit: DEFAULT_FOO implies that it's the fallback unless something specific was requested, which is not the case here as the value can be overridden. Would prefer to drop/replace the DEFAULT_ prefix.

Dropped the `DEFAULT` prefix.

> Wondering: instead of setting it for every package, could we do it in hosts/*.mk?

Reworked to set both in hosts/*.mk. Is this what you had in mind?

fanquake commented:
Android build failure here looks unrelated, however the macOS one seems to be an issue. Doesn't happen when building natively, only when cross-compiling qt:
```bash
Configuring qt...
Creating qmake...
make[1]: Entering directory '/tmp/cirrus-ci-build/depends/work/build/x86_64-apple-darwin18/qt/5.12.11-07ec867c621/qtbase/qmake'
make[1]: Leaving directory '/tmp/cirrus-ci-build/depends/work/build/x86_64-apple-darwin18/qt/5.12.11-07ec867c621/qtbase/qmake'
Info: creating super cache file /tmp/cirrus-ci-build/depends/work/build/x86_64-apple-darwin18/qt/5.12.11-07ec867c621/.qmake.super
Info: creating cache file /tmp/cirrus-ci-build/depends/work/build/x86_64-apple-darwin18/qt/5.12.11-07ec867c621/.qmake.cache
Info: creating stash file /tmp/cirrus-ci-build/depends/work/build/x86_64-apple-darwin18/qt/5.12.11-07ec867c621/.qmake.stash
Command line: -bindir /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/bin -c++std c++1z -confirm-license -hostprefix /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native -no-compile-examples -no-cups -no-egl -no-eglfs -no-freetype -no-gif -no-glib -no-icu -no-ico -no-iconv -no-kms -no-linuxfb -no-libjpeg -no-libproxy -no-libudev -no-mtdev -no-openssl -no-openvg -no-reduce-relocations -no-sctp -no-securetransport -no-sql-db2 -no-sql-ibase -no-sql-oci -no-sql-tds -no-sql-mysql -no-sql-odbc -no-sql-psql -no-sql-sqlite -no-sql-sqlite2 -no-system-proxies -no-use-gold-linker -nomake examples -nomake tests -nomake tools -opensource -pkg-config -prefix /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18 -qt-libpng -qt-pcre -qt-harfbuzz -qt-zlib -static -v -no-feature-bearermanagement -no-feature-colordialog -no-feature-commandlineparser -no-feature-concurrent -no-feature-dial -no-feature-fontcombobox -no-feature-ftp -no-feature-http -no-feature-image_heuristic_mask -no-feature-keysequenceedit -no-feature-lcdnumber -no-feature-networkdiskcache -no-feature-networkproxy -no-feature-pdf -no-feature-printdialog -no-feature-printer -no-feature-printpreviewdialog -no-feature-printpreviewwidget -no-feature-sessionmanager -no-feature-socks5 -no-feature-sql -no-feature-sqlmodel -no-feature-statemachine -no-feature-syntaxhighlighter -no-feature-textbrowser -no-feature-textodfwriter -no-feature-topleveldomain -no-feature-udpsocket -no-feature-undocommand -no-feature-undogroup -no-feature-undostack -no-feature-undoview -no-feature-vnc -no-feature-wizard -no-feature-xml -release -silent -no-dbus -no-opengl -pch -no-feature-corewlan QMAKE_MACOSX_DEPLOYMENT_TARGET=10.14 -xplatform macx-clang-linux -device-option MAC_SDK_PATH=/tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers -device-option MAC_SDK_VERSION=10.15.6 -device-option CROSS_COMPILE=x86_64-apple-darwin18- -device-option MAC_TARGET=x86_64-apple-darwin18 -device-option XCODE_VERSION=12.1
Project ERROR: Cannot run target compiler 'env -u C_INCLUDE_PATH -u CPLUS_INCLUDE_PATH -u OBJC_INCLUDE_PATH -u OBJCPLUS_INCLUDE_PATH -u CPATH -u LIBRARY_PATH /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/bin/clang++ --target=x86_64-apple-darwin18 -mmacosx-version-min=10.14 -B/tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/bin -mlinker-version=609 -isysroot/tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers -stdlib=libc++ -stdlib++-isystem/tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers/usr/include/c++/v1 -Xclang -internal-externc-isystem/tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/lib/clang/10.0.1/include -Xclang -internal-externc-isystem/tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers/usr/include'. Output:
===================
clang version 10.0.1 
Target: x86_64-apple-darwin18
Thread model: posix
InstalledDir: /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/bin
 "/tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/bin/clang" -cc1 -triple x86_64-apple-macosx10.14.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model pic -pic-level 2 -mthread-model posix -mframe-pointer=all -fno-rounding-math -masm-verbose -munwind-tables -target-sdk-version=10.15.6 -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -target-linker-version 609 -v -resource-dir /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/lib/clang/10.0.1 -isysroot /tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers -isysroot /tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers -I /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/include -I /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/include -internal-isystem /tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers/usr/include/c++/v1 -internal-isystem /tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers/usr/local/include -internal-isystem /tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/lib/clang/10.0.1/include -internal-externc-isystem /tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers/usr/include -O2 -std=c11 -fdeprecated-macro -fdebug-compilation-dir /tmp/cirrus-ci-build/depends/work/build/x86_64-apple-darwin18/qt/5.12.11-07ec867c621/qtbase/mkspecs/features -ferror-limit 19 -fmessage-length 0 -stack-protector 1 -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fobjc-runtime=macosx-10.14.0 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -vectorize-loops -vectorize-slp -internal-externc-isystem/tmp/cirrus-ci-build/depends/x86_64-apple-darwin18/native/lib/clang/10.0.1/include -internal-externc-isystem/tmp/cirrus-ci-build/depends/SDKs/Xcode-12.1-12A7403-extracted-SDK-with-libcxx-headers/usr/include -o /tmp/--eaf147.o -x c++ -
error: invalid argument '-std=c11' not allowed with 'C++'
===================
Maybe you forgot to setup the environment?
```

Looks like `clang` is invoked, and our cflags are added `-O2 -std=c11`, however we end up with `-x c++` on the end of our invocation when running some qt feature tests? Which results in Clang barfing. Will investigate.

fanquake commented:
Rebased on master, and added a potential fix for the macOS cross-compile issue. Which is to just not append `QMAKE_CFLAGS` to `QMAKE_CXXFLAGS` in the macOS configuration.

fanquake commented:
Updated both commits to document the new variables in the depends README.

dongcarl commented:
Looking clean! A few points of discussion:
1. Wondering if it'd make sense to somehow export the `CXX_STANDARD` in config.site
2. Wondering if we should be putting `-std` flags in `CFLAGS` or `CC`, for any other flag I would argue for `CFLAGS`, but `-std` seems quite fundamental, and if I understand correctly prior versions of autoconf would put `-std` flags into `CC` when requested: https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/C-Compiler.html

fanquake commented:
> Wondering if it'd make sense to somehow export the CXX_STANDARD in config.site

Potentially. Although if we did export it, and then passed it through to, for example, `ax_cxx_compile_stdcxx`, that script doesn't support detecting C++20 or later.

> Wondering if we should be putting -std flags in CFLAGS or CC

I don't have a particular opinion here.

fanquake commented:
Rebased. @theuni @dongcarl do either of you have any additional thoughts here? Will add new Guix hashes to PR description.

dongcarl commented on depends/README.md:
```diff
@@ -96,6 +96,8 @@ The following can be set when running make: `make FOO=bar`
 - `BASE_CACHE`: Built packages will be placed here
 - `SDK_PATH`: Path where SDKs can be found (used by macOS)
 - `FALLBACK_DOWNLOAD_PATH`: If a source file can't be fetched, try here before giving up
+- `C_STANDARD`: Set the C standard version used. Defaults to `C11`.
+- `CXX_STANDARD`: Set the C++ standard version used. Defaults to `C++17`.
```
```suggestion
- `C_STANDARD`: Set the C standard version used. Defaults to `c11`.
- `CXX_STANDARD`: Set the C++ standard version used. Defaults to `c++17`.
```

fanquake replied:
Fixed.

dongcarl reviewed (commented):
I think we added a few more hosts between my last review and this one. I'm also wondering if there's a good way to have this in `hosts/default.mk` somehow to stop worrying about new hosts, though `hosts/default.mk` does give me the scaries whenever I think about changing it.

dongcarl commented:
@fanquake 
> Not sure if we want to do that here, or a follow up after further discussion.

I don't mind at all what route we take. Let me know what you'd like to do!

fanquake commented:
@dongcarl I'd probably rather consolidate in a follow up, after more discussion, if you're happy to review here.

dongcarl commented:
Code Review ACK f7595f1354f4618436fdab232000dc152bff315a

Quite a simple change, good consistency improvement can be seen in the non-`depends/hosts` file diffs

laanwj commented:
Code review ACK f7595f1354f4618436fdab232000dc152bff315a
Changes look correct to me. I have run a succesful build, but was unable to conclude from the logging whether every package passes the correct flag to gcc/g++ (as a lot do not show the verbose build commands).
